state/protocol/badger/state.go

```diff
@@ -82,23 +82,30 @@ func Bootstrap(
 		return nil, fmt.Errorf("cannot bootstrap invalid root snapshot: %w", err)
 	}
 
-	err = operation.RetryOnConflictTx(db, transaction.Update, func(tx *transaction.Tx) error {
+	segment, err := root.SealingSegment()
+	if err != nil {
+		return nil, fmt.Errorf("could not get sealing segment: %w", err)
+	}
 
-		// 1) insert each block in the root chain segment
-		err = state.bootstrapSealingSegment(root)(tx)
-		if err != nil {
-			return fmt.Errorf("could not bootstrap sealing chain segment: %w", err)
-		}
+	// bootstrap the sealing segment step 1: insert segment execution results.
+	// execution results must be inserted first.
+	err = state.bootstrapSealingSegmentExeResults(segment)
+	if err != nil {
+		return nil, fmt.Errorf("could not bootstrap sealing chain segment execution results: %w", err)
+	}
 
-		segment, err := root.SealingSegment()
-		if err != nil {
-			return fmt.Errorf("could not get sealing segment: %w", err)
-		}
+	err = operation.RetryOnConflictTx(db, transaction.Update, func(tx *transaction.Tx) error {
 		// sealing segment is in ascending height order, so the tail is the
 		// oldest ancestor and head is the newest child in the segment
 		// TAIL <- ... <- HEAD
-		head := segment[len(segment)-1] // reference block of the snapshot
-		tail := segment[0]              // last sealed block
+		head := segment.Blocks[len(segment.Blocks)-1] // reference block of the snapshot
```

Comment:
It's better to define `head` and `tail` method on the Segment struct, otherwise, it's kind of `leaking` the knowledge.

Also I found `head` and `tail` is a bit confusing, they are usually used for labelling the position of a element in a list. And head usually is the first one, but here it's the last one in the slice.

I would suggest to just use `highest` and `lowest`, which implies their height. And we would know the `lowest` is the last sealed block.

Reply:
I agree